src/hpa.c

```diff
@@ -96,176 +89,319 @@ hpa_shard_init(hpa_shard_t *shard, hpa_t *hpa, edata_cache_t *edata_cache,
-	    size_goal);
-	malloc_mutex_unlock(tsdn, &hpa->mtx);
-	if (edata != NULL) {
-		malloc_mutex_unlock(tsdn, &hpa->grow_mtx);
-		edata_arena_ind_set(edata, shard->ind);
-		return edata;
-	}
+	return !edata_hugeified_get(ps)
+	    && edata_nfree_get(ps) < (HUGEPAGE / PAGE) * 5 / 100;
+}
 
+/* Returns true on error. */
+static void
+hpa_hugify(edata_t *ps) {
+	assert(edata_size_get(ps) == HUGEPAGE);
+	assert(edata_hugeified_get(ps));
+	bool err = pages_huge(edata_base_get(ps), HUGEPAGE);
 	/*
-	 * No such luck. We've dropped mtx, so other allocations can proceed
-	 * while we allocate the new extent.  We know no one else will grow in
-	 * the meantime, though, since we still hold grow_mtx.
+	 * Eat the error; even if the hugeification failed, it's still safe to
+	 * pretend it didn't (and would require extraordinary measures to
+	 * unhugify).
 	 */
-	size_t alloc_size;
-	pszind_t skip;
+	(void)err;
+}
 
-	size_t hugepage_goal_min = HUGEPAGE_CEILING(size_goal);
+static void
+hpa_dehugify(edata_t *ps) {
+	/* Purge, then dehugify while unbacked. */
+	pages_purge_forced(edata_addr_get(ps), HUGEPAGE);
+	pages_nohuge(edata_addr_get(ps), HUGEPAGE);
+	edata_hugeified_set(ps, false);
+}
 
-	err = exp_grow_size_prepare(&hpa->exp_grow, hugepage_goal_min,
-	    &alloc_size, &skip);
-	if (err) {
-		malloc_mutex_unlock(tsdn, &hpa->grow_mtx);
-		return NULL;
+static edata_t *
+hpa_grow(tsdn_t *tsdn, hpa_shard_t *shard) {
+	malloc_mutex_assert_owner(tsdn, &shard->grow_mtx);
+	edata_t *ps = NULL;
+
+	/* Is there address space waiting for reuse? */
+	malloc_mutex_assert_owner(tsdn, &shard->grow_mtx);
+	ps = edata_list_inactive_first(&shard->unused_slabs);
+	if (ps != NULL) {
+		edata_list_inactive_remove(&shard->unused_slabs, ps);
+		return ps;
+	}
+
+	/* Is eden a perfect fit? */
+	if (shard->eden != NULL && edata_size_get(shard->eden) == HUGEPAGE) {
+		ps = shard->eden;
+		shard->eden = NULL;
+		return ps;
 	}
-	alloc_size = HUGEPAGE_CEILING(alloc_size);
 
 	/*
-	 * Eventually, we need to think about this more systematically, and in
-	 * terms of extent hooks.  For now, though, we know we only care about
-	 * overcommitting systems, and we're not going to purge much.
+	 * We're about to try to allocate from eden by splitting.  If eden is
+	 * NULL, we have to allocate it too.  Otherwise, we just have to
+	 * allocate an edata_t for the new psset.
 	 */
-	bool commit = true;
-	void *addr = pages_map(NULL, alloc_size, HUGEPAGE, &commit);
-	if (addr == NULL) {
-		malloc_mutex_unlock(tsdn, &hpa->grow_mtx);
-		return NULL;
+	if (shard->eden == NULL) {
+		/*
+		 * During development, we're primarily concerned with systems
+		 * with overcommit.  Eventually, we should be more careful here.
+		 */
+		bool commit = true;
+		/* Allocate address space, bailing if we fail. */
+		void *new_eden = pages_map(NULL, HPA_EDEN_SIZE, HUGEPAGE,
+		    &commit);
+		if (new_eden == NULL) {
+			return NULL;
+		}
+		malloc_mutex_lock(tsdn, &shard->mtx);
```

Comment:
Added a comment. We drop the lock so that other allocations or frees can succeed while growing (though, the allocations must be smaller than the size of this alloc, except in the case of races in which an intervening dalloc frees up a longer free range). By analogy with the PAC: you don't hold the dirty ecache lock during an entire grow_retained call.